### ISSUE TITLE
Fix S2187: Rule should not raise False Positives for class inheriting class with tests

### ISSUE THREAD

**Issue by ValentijnMakkenze** [SonarSource/sonar-csharp#1196]
### Description

We are getting S2187 warnings for unit test classes that have no test directly but inherit tests from there base class.

### Repro steps

Our units tests usually follow the class inheritance of the classes that are being tested.
We then end up with some classes that have no tests of their own because the base class covers all functionality.

  [TestFixture]
  public class TestA
  {
    [Test]
    public void Test1()
    {
      // test something
    }
  }

  [TestFixture]
  public class TestB : TestA
  {
    // no test here but base class test is executed
  }

### Expected behavior

No warning.

### Actual behavior

Warning that class has no tests.

### Known workarounds

Add an additional test to the class.

### Related information

* SonarC# Version 6.8.2.4717
* Visual Studio Version 15.6.0 Preview 6.0

**Comment by Evangelink** [comment]
Hi @ValentijnMakkenze,

Thanks for the feedback. 

May I ask whether this example is a simplification of the real scenario or if it does mimic the real case? I cannot see the benefice of this `TestB` class. I could understand if `TestA` was `abstract` or defining a generic type provided by sub-classes but in this scenario of an empty `TestB` I fail to see the value.

**Comment by ValentijnMakkenze** [comment]
Hi Amaury,

I simplified it, TestA is indeed abstract in this case we have lots of simple codetables in the database and the CRUD test is in the abstract baseclass

  [TestFixture]
  public abstract class TestBusCodeBase<TBl, TBo, TTa, TDo, TTd>
    where TBl : BusBrkCodeNumberWithCodeList<TBl, TBo, TTa, TDo, TTd>
    where TBo : BusBrkCodeNumberWithCode<TBl, TBo, TTa, TDo, TTd>
    where TTa : TaBrkCodeNumberWithCode<TDo, TTd>, new()
    where TDo : DtoBrkCodeNumberWithCode, new()
    where TTd : TdBrkCodeNumberWithCode<TTd>, new()
  {
    [Test]
    public void TestCrud()
    {
      const string ExpectedCode = "XXXXX";
      const string ExpectedDescription = "Dummy code XXXXX";

      var target = BusCode<TBo, TTa, TDo>.NewRoot();
      target.Code = ExpectedCode;
      target.Description = ExpectedDescription;
      target = target.Save();
      Assert.That(target.Code, Is.EqualTo(ExpectedCode));
      Assert.That(target.Description, Is.EqualTo(ExpectedDescription));
      Assert.That(target.ToString(), Is.EqualTo(target.Description));
      target.Delete();
      target = target.Save();
      Assert.That(target, Is.Not.Null);
    }
  }

  [TestFixture]
  public class TestBusCAardfiliatie : TestBusCodeBase<BusCAardfiliatieList, BusCAardfiliatie, TaCAardfiliatie, DtoCAardfiliatie, TdCAardfiliatie>
  {
  }

Regards,

Valentijn

Van: Amaury Levé [mailto:notifications@github.com]
Verzonden: Wednesday, February 28, 2018 3:12 PM
Aan: SonarSource/sonar-csharp <sonar-csharp@noreply.github.com>
CC: Valentijn Makkenze <V.Makkenze@vicrea.nl>; Mention <mention@noreply.github.com>
Onderwerp: Re: [SonarSource/sonar-csharp] False positives for S2187 TestCases should contain tests (#1196)


Hi @ValentijnMakkenze<https://github.com/valentijnmakkenze>,

Thanks for the feedback.

May I ask whether this example is a simplification of the real scenario or if it does mimic the real case? I cannot see the benefice of this TestB class. I could understand if TestA was abstract or defining a generic type provided by sub-classes but in this scenario of an empty TestB I fail to see the value.

**Comment by Evangelink** [comment]
In this case that make sense! I'll have a look and see if I can reproduce the issue locally. If so I'll turn this ticket as False Positive ticket.

**Comment by sithmein** [comment]
It seems this issues either re-appeared or is not fully fixed yet. We are using Quarkus which allows to run tests in JVM and native mode. The tests are exactly the same but annotations on the test class designated them as either JVM or native test. Hence adding an abstract class with the actual test code makes perfect sense:

```
@NativeImageTest
class NativeHealthCheckTest extends AbstractHealthCheckTest {
 // empty
}
```
```
@QuarkusTest
class JVMHealthCheckTest extends AbstractHealthCheckTest {
 // empty
}
```

Yet in both cases Sonarqube 7.9 raises said violation which clearly doesn't make any sense.

**Comment by Evangelink** [comment]
@sithmein I might be missing something but my understanding is that you are talking about Java code and this is the dotnet repository. Could you please open a ticket in the forum?

Tagging @pavel-mikula-sonarsource and/or @andrei-epure-sonarsource

**Comment by sithmein** [comment]
Ah, sorry, you are right. Java it is.

**Comment by pavel-mikula-sonarsource** [comment]
@Shakevg Yes, please. This one is closed with a milestone on it.

**Comment by Shakevg** [comment]
> @Shakevg Yes, please. This one is closed with a milestone on it.

Submitted new issue https://github.com/SonarSource/sonar-dotnet/issues/5394
